zng/docs/spec.md

```diff
@@ -84,65 +83,10 @@ meta-records, and ZNG merely modernizes this original approach.
 The [`zq`](https://github.com/brimsec/zq) command-line tool provides a
 reference implementation of ZNG as it's described here, including the type
 system, error handling, etc., barring the exceptions
-described in the [alpha notice](#zng-specification)
+described in the [alpha notice](#note-this-specification-is-alpha-and-a-work-in-progress)
 at the top of this specification.
 
-## 2. The ZNG data model
-
-ZNG encodes a sequence of one or more typed data values to comprise a stream.
-The stream of values is interleaved with control messages
-that provide type definitions and other metadata.  The type of
-a particular data value is specified by its "type identifier", or type ID,
-which is an integer representing either a "primitive type" or a
-"container type".
-
-The ZNG type system comprises the standard set of primitive types like integers,
-floating point, strings, byte arrays, etc. as well as container types
-like records, arrays, and sets arranged from the primitive types.
-
-For example, a ZNG stream representing the single string "hello world"
-looks like this:
-```
-9:hello, world
-```
-Here, the type ID is the integer "9" representing the string type
-(defined in [Typedefs](#typedefs)) and the data value "hello, world"
-is an instance of the string type.
-
-ZNG gets more interesting when different data types are interleaved in the stream.
-For example,
-```
-9:hello, world
-4:42
-9:there's a fly in my soup!
-9:no, there isn't.
-4:3
-```
-where type ID 4 represents an integer.  This encoding represents the sequence of
-values:
-```
-"hello, world"
-42
-"there's a fly in my soup!"
-"no, there isn't."
-3
-```
-ZNG streams are often comprised as a sequence of records, which works well to provide
-an efficient representation of structured logs.  In this case, a new type ID is
-needed to define the schema for each distinct record.  To define a new
-type, the "#" syntax is used.  For example,
-logs from the open-source Zeek system might look like this
-```
-#24:record[_path:string,ts:time,uid:string,id:record[orig_h:ip,orig_p:port,resp_h:ip,resp_p:port]...
-#25:record[_path:string,ts:time,fuid:string,tx_hosts:set[ip]...
-24:[conn;1425565514.419939;CogZFI3py5JsFZGik;[192.168.1.1:;80/tcp;192.168.1.2;8080;]...
-25:[files;1425565514.419987;Fj8sRF1gdneMHN700d;[52.218.49.89;52.218.48.169;]...
-```
-Note that the value encoding need not refer to the field names and types as both are 
-completely captured by the type ID.  Values merely encode the value
-information consistent with the referenced type ID.
-
-## 3. ZNG Binary Format (ZNG)
+## 2. ZNG Binary Format (ZNG)
```

Comment:
I expect this may be controversial (cc: @mccanne, as I'm sure this text started with you in the very beginning) but @henridf and I have spoken offline a fair amount about it and I'm ready to advocate for removing Section 2 entirely. Read with fresh eyes, it looks like yet another relic from the ZSON days when things were text-only, so it was fair game to lead off with simple human-readable examples in hopes of pulling the reader along. However, in the time since then, not only do we now lead with binary ZNG as the more important piece, the examples that were in Section 2 that look like TZNG aren't even valid TZNG anymore, i.e. the `9` and `4` shown below are _not_ actually type IDs. To provide an accurate example we'd need to get into the details of tag integers vs. underlying type IDs as we do later in the doc where that feels more justified. If left as-is, we'd effectively be leading off with examples in pseudocode, and I don't think that serves the reader well.Since the other text in this section is already redundant with things covered elsewhere in the spec (Intro, TZNG Examples, etc.), I'm in favor of ripping off the band-aid and removing it.

Since records are the one "value message" that the `zq` reference implementation supports right now (as mentioned in the alpha notice up top) I _did_ put a Zeek example in Section 3.4 to take the place of the one that had been here.

Reply:
I think it's useful to describe the data model before defining the protocols that implement it, so I think we should keep this section.

Reply:
@nwt: Do you feel strongly enough that you're arguing for keeping the pseudocode examples as-is, or are you speaking up more in defense of the descriptive text outside the examples?

My first thoughts were to try and update the section so it could be kept, but as mentioned above, I found that difficult to do that while keeping the not-really-ZNG-and-not-really-TZNG examples. And once I started dropping examples, the remaining text didn't read as well by itself.

One approach I considered was to show the examples in a table (i.e. so there's no mistaking it for TZNG anymore) and be explicit about the fact that we're trying to present a human-readable representation of the soon-to-be-revealed binary format. After keeping the section's first two paragraphs intact as-is, I might take it in this direction:

---
...

For example, the following illustrates a ZNG value message for the single string "hello world" (he actual binary encoding is described in detail in the following section):

| Type ID | Value          |
|---------|------------|
| `9`        | hello, world |

Here, the type ID is the integer `9` representing the string type (defined in Typedefs) and the data value "hello, world" is an instance of the string type.

ZNG gets more interesting when different data types are interleaved in the stream. For example,

| Type ID | Value          |
|---------|------------|
| `9`        | hello, world |
| `4`        | 42               |
| `9`        | there's a fly in my soup! |
| `9`        | no, there's isn't. |
| `4`        | 3 |

where type ID `4` represents a 32-bit signed integer. This encoding represents the sequence of values that could be expressed in JSON as:

```
"hello, world"
42
"there's a fly in my soup!"
"no, there isn't."
3
```

ZNG streams are often comprised as a sequence of records, which works well to provide an efficient representation of structured logs. In this case, a new type defines the schema for each distinct record. For example, the following shows type bindings and values in TZNG for open source Zeek `weird` and `ftp` events:

```
#24:record[_path:string,ts:time,uid:bstring,id:record[orig_h:ip,orig_p:port,resp_h:ip,resp_p:port],name:bstring,addl:bstring,notice:bool,peer:bstring]
24:[weird;1521911720.600843;C1zOivgBT6dBmknqk;[10.47.1.152;49562;23.217.103.245;80;]TCP_ack_underflow_or_misorder;-;F;zeek;]
#25:record[_path:string,ts:time,uid:bstring,id:record[orig_h:ip,orig_p:port,resp_h:ip,resp_p:port],user:bstring,password:bstring,command:bstring,arg:bstring,mime_type:bstring,file_size:uint64,reply_code:uint64,reply_msg:bstring,data_channel:record[passive:bool,orig_h:ip,resp_h:ip,resp_p:port],fuid:bstring]
25:[ftp;1521911724.699488;ChkumY1k35TmZFL0V3;[10.164.94.120;45905;10.47.27.80;21;]anonymous;nessus@nessus.org;PASV;-;-;-;227;Entering Passive Mode (172,20,0,80,200,63).;[T;10.164.94.120;172.20.0.80;51263;]-;]
```

Note that the value encoding need not refer to the field names and types as both are completely captured by the type definition. Values merely encode the value information consistent with the referenced type.

---

That is, I'd also move the Zeek example back into this section, but be explicit in presenting it as TZNG, and it's valid TZNG in this case, so that feels right. In so doing I've also replaced the references to "type ID" since those are not accurate for this TZNG, instead speaking about "type definition" and "referenced type" more in the abstract.

Do you think this is an improvement?

Reply:
I'm speaking only in defense of the descriptive text and think the section is fine with or without examples. It's certainly dry without them, but I think it's OK for a specification to be dry.

If you do want to keep the examples, the tables are fine, but I think straight TZNG with minimal explanatory text would better. The table approach effectively introduces a third notation for the data model, and that's something I think we should avoid.

Reply:
@nwt: Yes, a "third notation" is precisely what I was trying to get away from, since what was there before was just that: It looked like TZNG, but it wasn't valid TZNG. You're correct that my new "third notation" wasn't great either. But I was nervous that I'd have difficulty crafting "TZNG with minimal explanatory text" since there's some devil in the details about TZNG vs. ZNG that only become fully apparent (to me, at least) after both formats are described in detail (e.g. "Type Bindings" in TZNG vs. "Typedefs" in binary ZNG).

I've pushed yet another commit that shows the examples as correct TZNG while using accurate wording & says just a bit about binary ZNG to nod toward the subtleties that will be explained later in the spec. Let me know if you think it's an improvement. I'd also love to hear from @henridf since he's been along for this ride. Thanks.

Reply:
> I've pushed yet another commit that shows the examples as correct TZNG while using accurate wording & says just a bit about binary ZNG to nod toward the subtleties that will be explained later in the spec.

I think this section better serves its informational purpose if we omit mention of differences between ZNG and TZNG.

Reply:
@nwt: Agreed. The "differences" text I had there before was my attempt to handle the resistance I was sensing to the removing the standalone `9:hello, world` that used to be there. Since it seems everyone is now on-board why I've needed to include type bindings in order to use human-readable TZNG in this section, it looks like I can remove those two sentences, which I've done by accepting your suggestion. I think I can finally merge!